frr: init at 7.5.1

woffs opened the pull request:
###### Motivation for this change

- FRRouting is a successor of no-longer maintained quagga (#120195)
- new dependency: libyang
- NixOS: old quagga service and test adapted to frr

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 124506` at 51173f32 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>frr</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

SuperSandro2000 commented on pkgs/servers/frr/libyang.nix:
```diff
@@ -0,0 +1,20 @@
+
+stdenv.mkDerivation rec {
+  pname = "libyang";
+  version = "1.0.225";
+  src = fetchFromGitHub {
+    owner = "CESNET";
+    repo = "libyang";
+    rev = "v${version}";
+    sha256 = "1f00j7jyszcz9zx41w3amm54sk6qq5m5lzkdkxg79pl1n1r69yr4";
+  };
+  nativeBuildInputs = [ cmake pkg-config ];
```
```suggestion
  version = "1.0.225";

  src = fetchFromGitHub {
    owner = "CESNET";
    repo = "libyang";
    rev = "v${version}";
    sha256 = "1f00j7jyszcz9zx41w3amm54sk6qq5m5lzkdkxg79pl1n1r69yr4";
  };

  nativeBuildInputs = [ cmake pkg-config ];
```

SuperSandro2000 commented on pkgs/servers/frr/libyang.nix:
```diff
@@ -0,0 +1,20 @@
+  nativeBuildInputs = [ cmake pkg-config ];
+  buildInputs = [ pcre ];
+  cmakeFlags = [ "-DENABLE_LYD_PRIV=ON" "-DCMAKE_BUILD_TYPE:String=Release" ];
+}
```
Missing meta. Why is this not a top-level entry?

woffs replied:
Just was not sure if it is worth to be a top-level entry.

woffs replied:
where to put it in the tree? `pkgs/development/tools/`?

woffs replied:
libraries, of course. done.

SuperSandro2000 commented on pkgs/servers/frr/default.nix:
```diff
@@ -0,0 +1,91 @@
+    "--enable-snmp"
+    "--enable-multipath=64"
+  ];
+
+  postInstall = ''
+    substituteInPlace $out/libexec/frr/frr-reload --replace /usr/lib/frr/ $out/libexec/frr/
+  '';
```
Can this be put into postPatch somehow?

woffs replied:
I think it can. Why is postPatch preferred?

woffs replied:
done.

SuperSandro2000 commented on pkgs/servers/frr/default.nix:
```diff
@@ -0,0 +1,91 @@
+    sha256 = "0lzsvi3kl9zcd4k04vc3363z9v2yrp7wc8bziv6w9h5fznh2vkxp";
+  };
+
+  buildInputs =
+    [ readline net-snmp c-ares json_c python3 libyang openssl czmq ]
+    ++ lib.optionals stdenv.isLinux [ libcap ];
```
```suggestion
  buildInputs = [ readline net-snmp c-ares json_c python3 libyang openssl czmq ]
    ++ lib.optionals stdenv.isLinux [ libcap ];
```

SuperSandro2000 reviewed (commented):
packages LGTM

higebu commented:
@aanderse @mweinelt Could you review this PR?
I think you can review this PR, because [gobgpd package initial release](https://github.com/NixOS/nixpkgs/pull/111316) was reviewed by @aanderse, @mweinelt and @SuperSandro2000 .
Thank you.

mweinelt reviewed (changes requested):
This needs to be split up into multiple commits. One for each package, module or test, so at least four commits.

mweinelt commented on nixos/modules/services/networking/frr.nix:
```diff
@@ -0,0 +1,186 @@
+            daemon = daemonName service;
+          in
+            nameValuePair daemon ({
+              wantedBy = [ "multi-user.target" ];
```
```suggestion
              wantedBy = [ "multi-user.target" ];
              after = [ "network.target" ];
```

Looks to be common to all services.

woffs replied:
"after" is already sensibly defined in the next paragraph.

mweinelt replied:
But it is common among all units defined, so might as well add it once here. Lists will merge anyway.

mweinelt commented on nixos/modules/services/networking/frr.nix:
```diff
@@ -0,0 +1,186 @@
+                  + optionalString (scfg.vtyListenAddress != "") " -A ${scfg.vtyListenAddress}"
+                  + optionalString (scfg.vtyListenPort != null) " -P ${toString scfg.vtyListenPort}";
+                ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+                Restart = "on-abort";
```
```suggestion
                Restart = "on-abnormal";
```

The upstream systemd unit uses that.

https://github.com/FRRouting/frr/blob/6f8a5e6157a88e42d2afd617832c8c3020f6df7e/tools/frr.service.in

There is probably more in there, that can and should be reused.

mweinelt commented on nixos/modules/services/networking/frr.nix:
```diff
@@ -0,0 +1,186 @@
+                  preStart = ''
+                    install -m 0755 -o frr -g frr -d /run/frr
+
+                    ${pkgs.iproute2}/bin/ip route flush proto zebra
```
Does this purge routes, that are not in the default table?

woffs replied:
I guess it removes obsolete routes from a unclean shutdown. It's inherited from the old quagga service.

mweinelt replied:
Right, but zebra could insert routes into just about any routing table, hence I'm left wondering if those get purged as well. Please check how upstream handles these leftover routes.

woffs replied:
seems to be obsolete, according to https://github.com/FRRouting/frr/commit/9ee2f5741ded52bef8ee20406257bdce56c5ea19

mweinelt commented on nixos/tests/frr.nix:
```diff
@@ -0,0 +1,96 @@
+              for gw in router1, router2:
+                  gw.wait_for_unit("ospfd")
+                  gw.wait_until_succeeds("vtysh -c 'show ip ospf neighbor' | grep Full")
+                  gw.wait_until_succeeds("vtysh -c 'show ip route' | grep '^O>'")
```
Not a user, but what does `0>` mean? Maybe this could be clearer?

mweinelt replied:
Yeah, it shows I'm not an FRR or Quagga user, more on the BIRD side.

mweinelt commented on nixos/tests/frr.nix:
```diff
@@ -0,0 +1,96 @@
+
+          with subtest("Test whether HTTP works"):
+              server.wait_for_unit("httpd")
+              client.succeed("curl --fail http://server/ >&2")
```
Testing HTTP is probably a bonus, after checking that ICMP works. But I'm really wondering if you really want to create a dependency in this test to the apache2 http server, seems unneccesarilly heavy.

woffs replied:
Removing the http test should be ok. Or use busybox? ...

mweinelt commented on pkgs/development/libraries/libyang/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+stdenv.mkDerivation rec {
+  pname = "libyang";
+  version = "1.0.225";
```
https://github.com/CESNET/libyang/releases/tag/v1.0.240

woffs replied:
thanks, will update

mweinelt commented on nixos/modules/services/networking/frr.nix:
```diff
@@ -0,0 +1,186 @@
+
+              serviceConfig = {
+                Type = "forking";
+                PIDFile = "/run/frr/${daemon}.pid";
```
Who creates `/run/frr`?

woffs replied:
apparently [libfrr](https://github.com/FRRouting/frr/blob/d8baf3db2d559fc13b3137ba9e3c6fa998f7e773/lib/libfrr.c#L721)

mweinelt replied:
We could probably just set `RuntimeDirectory=` for good measure.

woffs replied:
Maybe we could, but I'd rather not repair what's not broken.

woffs replied:
I found `install -m 0755 -o frr -g frr -d /run/frr` in preStart, this is ugly, will convert this into RuntimeDir.

woffs replied:
Or maybe use upstream `frrinit.sh`. Or think about it.

mweinelt replied:
To clarify:

- I don't think using their frrinit.sh init script is a good idea.
- Since multiple units use the RuntimeDirectory it's probably a good idea to set it up with systemd-tmpfiles

aanderse replied:
Or `RuntimeDirectoryPreserve`.

mweinelt reviewed (changes requested):
Looks pretty good already, a few recommendations and open questions.

mweinelt commented:
And since this introduces a new module we also require a changelog entry. Please see the new checkboxes in the pull request template and maybe update the top post accordingly.

https://github.com/NixOS/nixpkgs/blob/master/.github/PULL_REQUEST_TEMPLATE.md

aanderse commented on nixos/modules/services/networking/frr.nix:
```diff
@@ -0,0 +1,186 @@
+          in
+            nameValuePair daemon ({
+              wantedBy = [ "multi-user.target" ];
+              restartTriggers = [ (configFile service) ];
```
This is redundant and can be removed.

`restartTriggers` doesn't actually **do** anything... the intention of it is to introduce text which will be different every time you rebuild (because the path in the nix store is different) - this causes the service to be restarted because the text in the service changed. You already reference the configuration file elsewhere in the service file.

mweinelt replied:
Though maybe you want to have it reloaded, not restarted should the configuration change.

That's what `reloadIfChanged` is for.

aanderse replied:
Though if you enable `reloadIfChanged` your service won't restart when the binary changes... potentially introducing security vulnerabilities. It's a delicate balance on when to use that or not.

Regardless though, `restartTriggers` is not needed.

mweinelt replied:
Routing daemons are really stateful beasts. If handled wrong they interrupt production traffic, that's usually not a great spot to be in.

aanderse replied:
I'll be honest... I've been really busy and didn't even read what this software is... I figured you had it covered, but just wanted to mention the `restartTriggers` bit as most people don't understand the option (it's very complicated and "broken").

`reloadIfChanged` does sound good in that case!

:smile:

woffs replied:
I'll keep it default `restartIfChanged` for now, because this is what quagga was doing and I haven't heard of problems. Will remove restartTriggers, good point.

woffs committed:
libyang: init at 1.0.240

needed by frr

woffs committed:
frr: init at 7.5.1

- FRRouting is a successor of no-longer maintained quagga.
- new dependency: libyang

woffs committed:
nixos/frr: init

- old quagga service and test adapted to frr

woffs committed:
nixos/tests/frr: init

woffs commented:
I've split this up into 4 commits, removed apache2 from test and corrected some issues found by the reviewers, and rebased to master. Thanks for reviewing!

dminuoso commented on nixos/modules/services/networking/frr.nix:
```diff
@@ -0,0 +1,201 @@
+                  + optionalString (scfg.vtyListenAddress != "") " -A ${scfg.vtyListenAddress}"
+                  + optionalString (scfg.vtyListenPort != null) " -P ${toString scfg.vtyListenPort}";
+                ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+                Restart = "on-abnormal";
```
I'm not convinced this systemd structure is the right approach here.

- There's no proper supervision of unresponsive daemons. 
- Starting daemons must happen in a particular order, specifically zebra must be started first
- Restarting requires careful separation of daemons into phases.

Upstream frr comes with a systemd unit that simply runs watchfrr, handling all of those above points.

dminuoso replied:
Further, this systemd unit *must* reloadIfChanged, otherwise this tears down all routing protocol sessions, removes routes from RIB, etc...

woffs replied:
You're right, watchfrr should be used and this needs cleanup and rework. I don't know when I can do it yet.

mweinelt replied:
> There's no proper supervision of unresponsive daemons.

What does frr do in that case? I don't grok how `watchfrr` recognizes unresponsive daemons.

> Starting daemons must happen in a particular order, specifically zebra must be started first

That's easy, we let everything run after zebra. But I'm not sure what to do about all the other daemons. The dependencies are a list, not a tree. https://github.com/FRRouting/frr/blob/5af4491693676f7c52aeed3a981e6621a3a8bba4/tools/frrcommon.sh.in#L34-L38

> Restarting requires careful separation of daemons into phases.

Restarting with NixOS will be tough, since on a rebuild switch nixos will call `systemctl restart zebra.service ospfd.service bgpd.service` and systemd will not care about ordering. No idea how to resolve this.

woffs commented:
successor is #151946
